src/hooks/useOutsideClickDetector.ts

```diff
@@ -0,0 +1,34 @@
+import React, { useEffect } from 'react';
+
+/**
+ * A hook that detects clicking outside an element.
+ *
+ * @param {React.RefObject<HTMLElement>} ref
+ * @param {()=> void} onClickOutsideDetected
+ * @param {boolean} enableDetection
+ */
+const useOutsideClickDetector = (
+  ref: React.RefObject<HTMLElement>,
+  onClickOutsideDetected: () => void,
+  enableDetection: boolean,
+) => {
+  useEffect(() => {
+    const handleClickOutside = (event) => {
+      // if we click on an element inside the document that is not an inclusive descendant of the ref node.
+      if (ref.current && !ref.current.contains(event.target)) {
+        onClickOutsideDetected();
+      }
+    };
+    // no need to attach the listener if the parent component's visibility is controlled.
+    if (enableDetection) {
+      // Bind the event listener
+      document.addEventListener('mousedown', handleClickOutside);
+    }
+    return () => {
+      // Unbind the event listener on clean up
+      document.removeEventListener('mousedown', handleClickOutside);
+    };
+  }, [ref, onClickOutsideDetected, enableDetection]);
+};
+
+export default useOutsideClickDetector;
```

Comment:
Love this. Did you write or is it copied from somewhere? 

Also, can this work with multiple elements at the same time. Or can you only have a single global listener?

Reply:
No it's written from scratch :) . Also it works with multiple elements at once so that at any given time, only one dropdown will be visible because if you click at any verse's actions icon, it's considered you clicked outside of all other verse actions' icons which will cause them to close (if any is opened).